lib/ansible/module_utils/basic.py

```diff
@@ -2606,32 +2606,31 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
             strings on python3, use encoding=None to turn decoding to text off.
         '''
 
-        if isinstance(args, list):
-            if use_unsafe_shell:
-                args = " ".join([shlex_quote(x) for x in args])
-        elif isinstance(args, (binary_type, text_type)):
+        if isinstance(args, (binary_type, text_type)):
             # On python2.6 and below, shlex has problems with text type
             # On python3, shlex needs a text type.
             if PY2:
                 args = to_bytes(args, errors='surrogate_or_strict')
             elif PY3:
                 args = to_text(args, errors='surrogateescape')
             args = shlex.split(args)
-        else:
+        elif not isinstance(args, list):
             msg = "Argument 'args' to run_command must be list or string"
             self.fail_json(rc=257, cmd=args, msg=msg)
 
+        # expand shellisms
+        args = [os.path.expanduser(os.path.expandvars(x)) for x in args if x is not None]
```

Comment:
I don't think this is safe.  If someone is calling this function without unsafe_shell they probably have not quoted the arguments to prevent this sort of expansion.

Reply:
Okay, looking further down, I see that you're just moving this around though...

Reply:
it is a change in the end as we did not do this for stringified version .. but we probably should have.

Reply:
the shell itself would have done it before.  but might have done it slightly differently.

Reply:
so with last change to always stringify in unsafe, moving this back down into an else: